man/augtool.1
.\" Automatically generated by Pod::Man 4.14 (Pod::Simple 3.43)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "AUGTOOL 1"
.TH AUGTOOL 1 "2022-10-24" "Augeas 1.13.0" "Augeas"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
augtool \- inspect and modify configuration files
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
augtool [\s-1OPTIONS\s0] [\s-1COMMAND\s0]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Augeas is a configuration editing tool. It parses configuration files
in their native formats and transforms them into a tree. Configuration
changes are made by manipulating this tree and saving it back into
native config files.
.PP
augtool provides a command line interface to the generated tree. \s-1COMMAND\s0
can be a single command as described under \*(L"\s-1COMMANDS\*(R"\s0. When called with
no \s-1COMMAND,\s0 it reads commands from standard input until an end-of-file is
encountered.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-c\fR, \fB\-\-typecheck\fR" 4
.IX Item "-c, --typecheck"
Typecheck lenses. This can be very slow, and is therefore not done by
default, but is highly recommended during development.
.IP "\fB\-b\fR, \fB\-\-backup\fR" 4
.IX Item "-b, --backup"
When files are changed, preserve the originals in a file with extension
\&'.augsave'
.IP "\fB\-n\fR, \fB\-\-new\fR" 4
.IX Item "-n, --new"
Save changes in files with extension '.augnew', do not modify the original
files
.IP "\fB\-r\fR, \fB\-\-root\fR=\fI\s-1ROOT\s0\fR" 4
.IX Item "-r, --root=ROOT"
Use directory \s-1ROOT\s0 as the root of the filesystem. Takes precedence over a
root set with the \s-1AUGEAS_ROOT\s0 environment variable.
.IP "\fB\-I\fR, \fB\-\-include\fR=\fI\s-1DIR\s0\fR" 4
.IX Item "-I, --include=DIR"
Add \s-1DIR\s0 to the module loadpath. Can be given multiple times. The
directories set here are searched before any directories specified in the
\&\s-1AUGEAS_LENS_LIB\s0 environment variable, and before the default directories
\&\fI/usr/share/augeas/lenses\fR and \fI/usr/share/augeas/lenses/dist\fR.
.IP "\fB\-t\fR, \fB\-\-transform\fR=\fI\s-1XFM\s0\fR" 4
.IX Item "-t, --transform=XFM"
Add a file transform; uses the 'transform' command syntax,
e.g. \f(CW\*(C`\-t \*(AqFstab incl /etc/fstab.bak\*(Aq\*(C'\fR.
.IP "\fB\-l\fR, \fB\-\-load\-file\fR=\fI\s-1FILE\s0\fR" 4
.IX Item "-l, --load-file=FILE"
Load an individual \s-1FILE\s0 into the tree. The lens to use is determined
automatically (based on autoload information in the lenses) and will be the
same that is used for this file when the entire tree is loaded. The option
can be specified multiple times to load several files, e.g. \f(CW\*(C`\-l /etc/fstab
\&\-l /etc/hosts\*(C'\fR. This lens implies \f(CW\*(C`\-\-noload\*(C'\fR so that only the files
specified with this option will be loaded.
.IP "\fB\-f\fR, \fB\-\-file\fR=\fI\s-1FILE\s0\fR" 4
.IX Item "-f, --file=FILE"
Read commands from \s-1FILE.\s0
.IP "\fB\-i\fR, \fB\-\-interactive\fR" 4
.IX Item "-i, --interactive"
Read commands from the terminal. When combined with \fB\-f\fR or redirection of
stdin, drop into an interactive session after executing the commands from
the file.
.IP "\fB\-e\fR, \fB\-\-echo\fR" 4
.IX Item "-e, --echo"

NEWS
1.14.0 - 2022-12-07
  - General changes/additions
    * Update submodule gnulib to 2f7479a16a3395f1429c7795f10c5d19b9b4453e (#781)
    * Add bash-completion for augtool, augmatch, augprint (#783) George Hansper
    * Fix: Allow values to contain arbitrary unbalanced square brackets (#782) George Hansper
    * Add package bash to build stage in Dockerfile (#776) George Hansper
    * Add augprint tool for creating idempotent augtool scripts (#752) George Hansper
    * Replace deprecated 'security_context_t' with 'char *' (#747) Leo-Schmit
    * src/syntax.c: Fix whitespace which confuses static checkers (#725) rwmjones
    * README.md: Add oss-fuzz status badge (#702) Sergey Nizovtsev
    * Package augmatch, too (#688) oleksandriegorov
    * Add Github workflow to create releases with complete source tarballs (#744) Hilko Bengen

  - Lens changes/additions
    * Resolv: add option trust-ad (#784) George Hansper
    * Sos: new lens for /etc/sos/sos.conf (based on IniFile)
           (#779) George Hansper
    * Pg_Hba: unquoted auth-method may contain hyphens (#777) George Hansper
    * Sysctl: Allow keys to contain * and : and / characters (#755) M Filka
    * Semanage: Fix parsing of ignoredirs (#758) Richard W M Jones
    * Systemd: allow empty quoted environment variable values
               (#757) Michal Vasko
    * Systemd: allow values starting with whitespaces for Exec* and
               Environment service entries. (#757) Michal Vasko
    * Toml: workaround to allow writing toml files (#742) Richard
    * Kdump: parse "auto_reset_crashkernel" (#754) Laszlo Ersek
    * Keepalived: add parameters notify_stop and notify_deleted (#749) Adam Bambuch
    * Chrony: add new directives and options (#745) Miroslav Lichvar
    * Redis: Allow redis lens to set 'SAVE ""' as a valid option
             (#738) Mitch Hagstrand
    * ClamAV: update ClamAV lens to autoload /etc/clamav/*.conf
              (#748) Guillaume Ross
    * AuthselectPam: new lens for /etc/authselect/custom/*/*-auth and
          /etc/authselect/custom/*/postlogin (#743) Heston Snodgrass
    * Sshd: Parse GSSAPIKexAlgorithms PubkeyAcceptedKeyTypes CASignatureAlgorithms
            as comma-seperated lists instead of simple strings
            (#721) Edward Garbade
    * Yum: Add additional unit tests (#677) Pat Riehecky
    * Cockpit: new lens for /etc/cockpit/cockpit.conf (#675) Pat Riehecky

1.13.0 - 2021-10-15
  - General changes/additions
    * Add Dockerfile (Nicolas Gif) (Issue #650)
    * augtool: Improved readline integration to handle quoting issues
               (Pino Toscano)
    * typechecker: Allow including '/' in keys and labels. Thanks to
                   felixdoerre for pointing out that this restriction was
                   unnecessary. See issue #668 for the discussion.
    * Add function modified() to select nodes which are marked as dirty
      (George Hansper) (Issue #691)
    * Add CLI command 'preview' and API 'aug_preview' to preview file contents
      (George Hansper) (#690)
    * Add "else" operator to augeas path-filter expressions (priority selector)
      (George Hansper) (#692)
    * Add new axis 'seq' to allow /path/seq::*[expr] to match and create numeric
      nodes, as idempotent alternative to /path/*[expr] (George Hansper) (#706)

  - Lens changes/additions
    * Authinfo2: new lens to parse Authinfo2 format (Nicolas Gif) (Issue #649)
    * Chrony: add new options (Miroslav Lichvar) (Issue #698)

src/augtool.c
/*
 * augtool.c:
 *
 * Copyright (C) 2007-2016 David Lutterkort
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
 *
 * Author: David Lutterkort <dlutter@redhat.com>
 */

#include <config.h>
#include "augeas.h"
#include "internal.h"
#include "safe-alloc.h"

#include <readline/readline.h>
#include <readline/history.h>
#include <argz.h>
#include <getopt.h>
#include <limits.h>
#include <ctype.h>
#include <locale.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <pwd.h>
#include <stdarg.h>
#include <sys/time.h>

/* Global variables */

static augeas *aug = NULL;
static const char *const progname = "augtool";
static unsigned int flags = AUG_NONE;
const char *root = NULL;
char *loadpath = NULL;
char *transforms = NULL;
char *loadonly = NULL;
size_t transformslen = 0;
size_t loadonlylen = 0;
const char *inputfile = NULL;
int echo_commands = 0; /* Gets also changed in main_loop */
bool print_version = false;
bool auto_save = false;
bool interactive = false;
bool timing = false;
/* History file is ~/.augeas/history */
char *history_file = NULL;

#define AUGTOOL_PROMPT "augtool> "

/*
 * General utilities
 */

/* Private copy of xasprintf from internal to avoid Multiple definition in
 * static builds.
 */
static int _xasprintf(char **strp, const char *format, ...)
{
    va_list args;
    int result;

    va_start(args, format);
    result = vasprintf(strp, format, args);
    va_end(args);
    if (result < 0)
        *strp = NULL;
    return result;
}

static int child_count(const char *path)
{
    char *pat = NULL;
    int r;

    if (path[strlen(path) - 1] == SEP)
        r = asprintf(&pat, "%s*", path);
    else
        r = asprintf(&pat, "%s/*", path);
    if (r < 0)
        return -1;
    r = aug_match(aug, pat, NULL);
    free(pat);

    return r;
}

static char *readline_path_generator(const char *text, int state)
{
    static int current = 0;
    static char **children = NULL;
    static int nchildren = 0;
    static char *ctx = NULL;

    char *end = strrchr(text, SEP);
    if (end != NULL)
        end += 1;

    if (state == 0)
    {
        char *path;
        if (end == NULL)
        {
            if ((path = strdup("*")) == NULL)
                return NULL;
        }
        else
        {
            if (ALLOC_N(path, end - text + 2) < 0)
                return NULL;
            strncpy(path, text, end - text);
            strcat(path, "*");
        }

        for (; current < nchildren; current++)
            free((void *)children[current]);
        free((void *)children);
        nchildren = aug_match(aug, path, &children);
        current = 0;

        ctx = NULL;
        if (path[0] != SEP)
            aug_get(aug, AUGEAS_CONTEXT, (const char **)&ctx);

        free(path);
    }

    if (end == NULL)
        end = (char *)text;

    while (current < nchildren)
    {
        char *child = children[current];
        current += 1;

        char *chend = strrchr(child, SEP) + 1;
        if (STREQLEN(chend, end, strlen(end)))
        {
            if (child_count(child) > 0)
            {
                char *c = realloc(child, strlen(child) + 2);
                if (c == NULL)
                {
                    free(child);
                    return NULL;
                }
                child = c;
                strcat(child, "/");
            }

            /* strip off context if the user didn't give it */
            if (ctx != NULL)
            {
                int ctxidx = strlen(ctx);
                if (child[ctxidx] == SEP)
                    ctxidx++;
                char *c = strdup(&child[ctxidx]);
                free(child);
                if (c == NULL)
                    return NULL;
                child = c;
            }

            rl_filename_completion_desired = 1;
            rl_completion_append_character = '\0';
            return child;
        }
        else
        {
            free(child);
        }
    }
    return NULL;
}

static char *readline_command_generator(const char *text, int state)
{
    // FIXME: expose somewhere under /augeas
    static const char *const commands[] = {
        "quit", "clear", "defnode", "defvar",
        "get", "label", "ins", "load", "ls", "match",
        "mv", "cp", "rename", "print", "dump-xml", "dump-json", "rm", "save", "set", "setm",
        "clearm", "span", "store", "retrieve", "transform", "load-file",
        "help", "touch", "insert", "move", "copy", "errors", "source", "context",
        "info", "count", "preview",
        NULL};

    static int current = 0;
    const char *name;

    if (state == 0)
        current = 0;

    rl_completion_append_character = ' ';
    while ((name = commands[current]) != NULL)
    {
        current += 1;
        if (STREQLEN(text, name, strlen(text)))
            return strdup(name);
    }
    return NULL;
}

#ifndef HAVE_RL_COMPLETION_MATCHES
typedef char *rl_compentry_func_t(const char *, int);
static char **rl_completion_matches(ATTRIBUTE_UNUSED const char *text,
                                    ATTRIBUTE_UNUSED rl_compentry_func_t *func)
{
    return NULL;
}
#endif

#ifndef HAVE_RL_CRLF
static int rl_crlf(void)
{
    if (rl_outstream != NULL)
        putc('\n', rl_outstream);
    return 0;
}
#endif

#ifndef HAVE_RL_REPLACE_LINE
static void rl_replace_line(ATTRIBUTE_UNUSED const char *text,
                            ATTRIBUTE_UNUSED int clear_undo)
{
    return;
}
#endif

static char **readline_completion(const char *text, int start,
                                  ATTRIBUTE_UNUSED int end)
{
    if (start == 0)
        return rl_completion_matches(text, readline_command_generator);
    else
        return rl_completion_matches(text, readline_path_generator);

    return NULL;
}

static char *get_home_dir(uid_t uid)
{
    char *strbuf;
    char *result;
    struct passwd pwbuf;
    struct passwd *pw = NULL;
    long val = sysconf(_SC_GETPW_R_SIZE_MAX);

    if (val < 0)
    {
        // The libc won't tell us how big a buffer to reserve.
        // Let's hope that 16k is enough (it really should be).
        val = 16 * 1024;
    }

    size_t strbuflen = (size_t)val;

    if (ALLOC_N(strbuf, strbuflen) < 0)
        return NULL;

    if (getpwuid_r(uid, &pwbuf, strbuf, strbuflen, &pw) != 0 || pw == NULL)
    {
        free(strbuf);

        // Try to get the user's home dir from the environment
        char *env = getenv("HOME");
        if (env != NULL)
        {
            return strdup(env);
        }
        return NULL;
    }

    result = strdup(pw->pw_dir);

    free(strbuf);

    return result;
}

/* Inspired from:
 * https://thoughtbot.com/blog/tab-completion-in-gnu-readline
 */
static int quote_detector(char *str, int index)
{
    return index > 0 && str[index - 1] == '\\' && quote_detector(str, index - 1) == 0;
}

static void readline_init(void)
{
    rl_readline_name = "augtool";
    rl_attempted_completion_function = readline_completion;
    rl_completion_entry_function = readline_path_generator;
    rl_completer_quote_characters = "\"'";
    rl_completer_word_break_characters = (char *)" ";
    rl_char_is_quoted_p = &quote_detector;

    /* Set up persistent history */
    char *home_dir = get_home_dir(getuid());
    char *history_dir = NULL;

    if (home_dir == NULL)
        goto done;

    if (_xasprintf(&history_dir, "%s/.augeas", home_dir) < 0)
        goto done;

    if (mkdir(history_dir, 0755) < 0 && errno != EEXIST)
        goto done;

    if (_xasprintf(&history_file, "%s/history", history_dir) < 0)
        goto done;

    stifle_history(500);

    read_history(history_file);

done:
    free(home_dir);
    free(history_dir);
}

__attribute__((noreturn)) static void help(void)
{
    fprintf(stderr, "Usage: %s [OPTIONS] [COMMAND]\n", progname);
    fprintf(stderr, "Load the Augeas tree and modify it. If no COMMAND is given, run interactively\n");
    fprintf(stderr, "Run '%s help' to get a list of possible commands.\n",
            progname);
    fprintf(stderr, "\nOptions:\n\n");
    fprintf(stderr, "  -c, --typecheck        typecheck lenses\n");
    fprintf(stderr, "  -b, --backup           preserve originals of modified files with\n"
                    "                         extension '.augsave'\n");
    fprintf(stderr, "  -n, --new              save changes in files with extension '.augnew',\n"
                    "                         leave original unchanged\n");
    fprintf(stderr, "  -r, --root ROOT        use ROOT as the root of the filesystem\n");
    fprintf(stderr, "  -I, --include DIR      search DIR for modules; can be given multiple times\n");
    fprintf(stderr, "  -t, --transform XFM    add a file transform; uses the 'transform' command\n"
                    "                         syntax, e.g. -t 'Fstab incl /etc/fstab.bak'\n");
    fprintf(stderr, "  -l, --load-file FILE   load individual FILE in the tree\n");
    fprintf(stderr, "  -e, --echo             echo commands when reading from a file\n");
    fprintf(stderr, "  -f, --file FILE        read commands from FILE\n");
    fprintf(stderr, "  -s, --autosave         automatically save at the end of instructions\n");
    fprintf(stderr, "  -i, --interactive      run an interactive shell after evaluating\n"
                    "                         the commands in STDIN and FILE\n");
    fprintf(stderr, "  -S, --nostdinc         do not search the builtin default directories\n"
                    "                         for modules\n");
    fprintf(stderr, "  -L, --noload           do not load any files into the tree on startup\n");
    fprintf(stderr, "  -A, --noautoload       do not autoload modules from the search path\n");
    fprintf(stderr, "  --span                 load span positions for nodes related to a file\n");
    fprintf(stderr, "  --timing               after executing each command, show how long it took\n");
    fprintf(stderr, "  --version              print version information and exit.\n");

    exit(EXIT_FAILURE);
}

static void parse_opts(int argc, char **argv)
{
    int opt;
    size_t loadpathlen = 0;
    enum
    {
        VAL_VERSION = CHAR_MAX + 1,
        VAL_SPAN = VAL_VERSION + 1,
        VAL_TIMING = VAL_SPAN + 1
    };
    struct option options[] = {
        {"help", 0, 0, 'h'},
        {"typecheck", 0, 0, 'c'},
        {"backup", 0, 0, 'b'},
        {"new", 0, 0, 'n'},
        {"root", 1, 0, 'r'},
        {"include", 1, 0, 'I'},
        {"transform", 1, 0, 't'},
        {"load-file", 1, 0, 'l'},
        {"echo", 0, 0, 'e'},
        {"file", 1, 0, 'f'},
        {"autosave", 0, 0, 's'},
        {"interactive", 0, 0, 'i'},
        {"nostdinc", 0, 0, 'S'},
        {"noload", 0, 0, 'L'},
        {"noautoload", 0, 0, 'A'},
        {"span", 0, 0, VAL_SPAN},
        {"timing", 0, 0, VAL_TIMING},
        {"version", 0, 0, VAL_VERSION},
        {0, 0, 0, 0}};
    int idx;

    while ((opt = getopt_long(argc, argv, "hnbcr:I:t:l:ef:siSLA", options, &idx)) != -1)
    {
        switch (opt)
        {
        case 'c':
            flags |= AUG_TYPE_CHECK;
            break;
        case 'b':
            flags |= AUG_SAVE_BACKUP;
            break;
        case 'n':
            flags |= AUG_SAVE_NEWFILE;
            break;
        case 'h':
            help();
            break;
        case 'r':
            root = optarg;
            break;
        case 'I':
            argz_add(&loadpath, &loadpathlen, optarg);
            break;
        case 't':
            argz_add(&transforms, &transformslen, optarg);
            break;
        case 'l':
            // --load-file implies --noload
            flags |= AUG_NO_LOAD;
            argz_add(&loadonly, &loadonlylen, optarg);
            break;
        case 'e':
            echo_commands = 1;
            break;
        case 'f':
            inputfile = optarg;
            break;
        case 's':
            auto_save = true;
            break;
        case 'i':
            interactive = true;
            break;
        case 'S':
            flags |= AUG_NO_STDINC;
            break;
        case 'L':
            flags |= AUG_NO_LOAD;
            break;
        case 'A':
            flags |= AUG_NO_MODL_AUTOLOAD;
            break;
        case VAL_VERSION:
            flags |= AUG_NO_MODL_AUTOLOAD;
            print_version = true;
            break;
        case VAL_SPAN:
            flags |= AUG_ENABLE_SPAN;
            break;
        case VAL_TIMING:
            timing = true;
            break;
        default:
            fprintf(stderr, "Try '%s --help' for more information.\n",
                    progname);
            exit(EXIT_FAILURE);
            break;
        }
    }
    argz_stringify(loadpath, loadpathlen, PATH_SEP_CHAR);
}

static void print_version_info(void)
{
    const char *version;
    int r;

    r = aug_get(aug, "/augeas/version", &version);
    if (r != 1)
        goto error;

    fprintf(stderr, "augtool %s <http://augeas.net/>\n", version);
    fprintf(stderr, "Copyright (C) 2007-2016 David Lutterkort\n");
    fprintf(stderr, "License LGPLv2+: GNU LGPL version 2.1 or later\n");
    fprintf(stderr, "                 <http://www.gnu.org/licenses/lgpl-2.1.html>\n");
    fprintf(stderr, "This is free software: you are free to change and redistribute it.\n");
    fprintf(stderr, "There is NO WARRANTY, to the extent permitted by law.\n\n");
    fprintf(stderr, "Written by David Lutterkort\n");
    return;
error:
    fprintf(stderr, "Something went terribly wrong internally - please file a bug\n");
}

static void print_time_taken(const struct timeval *start,
                             const struct timeval *stop)
{
    time_t elapsed = (stop->tv_sec - start->tv_sec) * 1000 + (stop->tv_usec - start->tv_usec) / 1000;
    printf("Time: %ld ms\n", elapsed);
}

static int run_command(const char *line, bool with_timing)
{
    int result;
    struct timeval stop, start;

    gettimeofday(&start, NULL);
    result = aug_srun(aug, stdout, line);
    gettimeofday(&stop, NULL);
    if (with_timing && result >= 0)
    {
        print_time_taken(&start, &stop);
    }

    if (isatty(fileno(stdin)))
        add_history(line);
    return result;
}

static void print_aug_error(void)
{
    if (aug_error(aug) == AUG_ENOMEM)
    {
        fprintf(stderr, "Out of memory.\n");
        return;
    }
    if (aug_error(aug) != AUG_NOERROR)
    {
        fprintf(stderr, "error: %s\n", aug_error_message(aug));
        if (aug_error_minor_message(aug) != NULL)
            fprintf(stderr, "error: %s\n",
                    aug_error_minor_message(aug));
        if (aug_error_details(aug) != NULL)
        {
            fputs(aug_error_details(aug), stderr);
            fprintf(stderr, "\n");
        }
    }
}

static void sigint_handler(ATTRIBUTE_UNUSED int signum)
{
    // Cancel the current line of input, along with undo info for that line.
    rl_replace_line("", 1);

    // Move the cursor to the next screen line, then force a re-display.
    rl_crlf();
    rl_forced_update_display();
}

static void install_signal_handlers(void)
{
    // On Ctrl-C, cancel the current line (rather than exit the program).
    struct sigaction sigint_action;
    MEMZERO(&sigint_action, 1);
    sigint_action.sa_handler = sigint_handler;
    sigemptyset(&sigint_action.sa_mask);
    sigint_action.sa_flags = 0;
    sigaction(SIGINT, &sigint_action, NULL);
}

static int main_loop(void)
{
    char *line = NULL;
    int ret = 0;
    char inputline[128];
    int code;
    bool end_reached = false;
    bool get_line = true;
    bool in_interactive = false;

    if (inputfile)
    {
        if (freopen(inputfile, "r", stdin) == NULL)
        {
            char *msg = NULL;
            if (asprintf(&msg, "Failed to open %s", inputfile) < 0)
                perror("Failed to open input file");
            else
                perror(msg);
            return -1;
        }
    }

    install_signal_handlers();

    // make readline silent by default
    echo_commands = echo_commands || isatty(fileno(stdin));
    if (echo_commands)
        rl_outstream = NULL;
    else
        rl_outstream = fopen("/dev/null", "w");

    while (1)
    {
        if (get_line)
        {
            line = readline(AUGTOOL_PROMPT);
        }
        else
        {
            line = NULL;
        }

        if (line == NULL)
        {
            if (!isatty(fileno(stdin)) && interactive && !in_interactive)
            {
                in_interactive = true;
                if (echo_commands)
                    printf("\n");
                echo_commands = true;

                // reopen in stream
                if (freopen("/dev/tty", "r", stdin) == NULL)
                {
                    perror("Failed to open terminal for reading");
                    return -1;
                }
                rl_instream = stdin;

                // reopen stdout and stream to a tty if originally silenced or
                // not connected to a tty, for full interactive mode
                if (rl_outstream == NULL || !isatty(fileno(rl_outstream)))
                {
                    if (rl_outstream != NULL)
                    {
                        fclose(rl_outstream);
                        rl_outstream = NULL;
                    }
                    if (freopen("/dev/tty", "w", stdout) == NULL)
                    {
                        perror("Failed to reopen stdout");
                        return -1;
                    }
                    rl_outstream = stdout;
                }
                continue;
            }

            if (auto_save)
            {
                strncpy(inputline, "save", sizeof(inputline));
                line = inputline;
                if (echo_commands)
                    printf("%s\n", line);
                auto_save = false;
            }
            else
            {
                end_reached = true;
            }
            get_line = false;
        }

        if (end_reached)
        {
            if (echo_commands)
                printf("\n");
            return ret;
        }

        if (*line == '\0' || *line == '#')
        {
            free(line);
            continue;
        }

        code = run_command(line, timing);
        if (code == -2)
        {
            free(line);
            return ret;
        }

        if (code < 0)
        {
            ret = -1;
            print_aug_error();
        }

        if (line != inputline)
            free(line);
    }
}

static int run_args(int argc, char **argv)
{
    size_t len = 0;
    char *line = NULL;
    int code;

    for (int i = 0; i < argc; i++)
        len += strlen(argv[i]) + 1;
    if (ALLOC_N(line, len + 1) < 0)
        return -1;
    for (int i = 0; i < argc; i++)
    {
        strcat(line, argv[i]);
        strcat(line, " ");
    }
    if (echo_commands)
        printf("%s%s\n", AUGTOOL_PROMPT, line);
    code = run_command(line, timing);
    free(line);
    if (code >= 0 && auto_save)
        if (echo_commands)
            printf("%ssave\n", AUGTOOL_PROMPT);
    code = run_command("save", false);

    if (code < 0)
    {
        code = -1;
        print_aug_error();
    }
    return (code >= 0 || code == -2) ? 0 : -1;
}

static void add_transforms(char *ts, size_t tslen)
{
    char *command;
    int r;
    char *t = NULL;
    bool added_transform = false;

    while ((t = argz_next(ts, tslen, t)))
    {
        r = _xasprintf(&command, "transform %s", t);
        if (r < 0)
            fprintf(stderr, "error: Failed to add transform %s: could not allocate memory\n", t);

        r = aug_srun(aug, stdout, command);
        if (r < 0)
            fprintf(stderr, "error: Failed to add transform %s: %s\n", t, aug_error_message(aug));

        free(command);
        added_transform = true;
    }

    if (added_transform)
    {
        r = aug_load(aug);
        if (r < 0)
            fprintf(stderr, "error: Failed to load with new transforms: %s\n", aug_error_message(aug));
    }
}

static void load_files(char *ts, size_t tslen)
{
    char *command;
    int r;
    char *t = NULL;

    while ((t = argz_next(ts, tslen, t)))
    {
        r = _xasprintf(&command, "load-file %s", t);
        if (r < 0)
            fprintf(stderr, "error: Failed to load file %s: could not allocate memory\n", t);

        r = aug_srun(aug, stdout, command);
        if (r < 0)
            fprintf(stderr, "error: Failed to load file %s: %s\n", t, aug_error_message(aug));

        free(command);
    }
}

int main(int argc, char **argv)
{
    int r;
    struct timeval start, stop;

    setlocale(LC_ALL, "");

    parse_opts(argc, argv);

    if (timing)
    {
        printf("Initializing augeas ... ");
        fflush(stdout);
    }
    gettimeofday(&start, NULL);

    aug = aug_init(root, loadpath, flags | AUG_NO_ERR_CLOSE);

    gettimeofday(&stop, NULL);
    if (timing)
    {
        printf("done\n");
        print_time_taken(&start, &stop);
    }

    if (aug == NULL || aug_error(aug) != AUG_NOERROR)
    {
        fprintf(stderr, "Failed to initialize Augeas\n");
        if (aug != NULL)
            print_aug_error();
        exit(EXIT_FAILURE);
    }
    load_files(loadonly, loadonlylen);
    add_transforms(transforms, transformslen);
    if (print_version)
    {
        print_version_info();
        return EXIT_SUCCESS;
    }
    readline_init();
    if (optind < argc)
    {
        // Accept one command from the command line
        r = run_args(argc - optind, argv + optind);
    }
    else
    {
        r = main_loop();
    }
    if (history_file != NULL)
        write_history(history_file);

    aug_close(aug);
    return r == 0 ? EXIT_SUCCESS : EXIT_FAILURE;
}

/*
 * Local variables:
 *  indent-tabs-mode: nil
 *  c-indent-level: 4
 *  c-basic-offset: 4
 *  tab-width: 4
 * End:
 */
